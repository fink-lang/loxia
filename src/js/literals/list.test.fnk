{fink2js} = import '../../testing/generate.fnk'
{skip, describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'


describe 'list', fn:
  it 'compiles', fn:
    expect
      fink2js '
        array1 = []
        array2 = [1]
        array3 = [1, 2]
        array4 = [1, , 2]
        array5 = list:
          a + 1 + 45 +
            b + c
          [1, 2]
          (3 + 3) * 2
      '
      to_match_snapshot


  it 'compiles spread', fn:
    expect
      fink2js '
        foo = [...b]
        bar = [1, 2, ...b]
        spam = [1, 2, ...b, 123]
      '
      to_match_snapshot

    expect
      fink2js '
        foo = [1, 2, 3]
        bar = [...foo, 4, 5, 6]
        spam = [7]
        shrub = [...bar, ...spam]
        ni = [...nu, na]
        ham = [...nu, ...na]
      '
      to_match_snapshot


  skip.it 'compiles as partial', fn:
    expect
      fink2js '
        [1, ...?, 9]
        [?, ?]
      '
      to_match_snapshot


describe 'unpacking list', fn:
  it 'destructures simple', fn:
    expect
      fink2js '
        [a, [b, c]] = [1, [2, 3]]
        foo = [b, a]
        [z] = [1 + 2, 3 * 4]
        [x, y] = foo
        out = [a, b, c, foo, z, x, y]
      '
      to_match_snapshot

  it 'destructures empty', fn:
    expect
      fink2js '
        [a,  , b] = foo
        out = [a, b]
      '
      to_match_snapshot

  it 'destructures spread', fn:
    expect
      fink2js '
        [a, ..., b] = foo
        out = [a, b]
      '
      to_match_snapshot

    expect
      fink2js '
        [a, ..., b, c] = foo
        out = [a, b, c]
      '
      to_match_snapshot

    expect
      fink2js '
        [a, ...b] = foo
        out = [a, ...b]
      '
      to_match_snapshot

    expect
      fink2js '
        [a, ...b, c] = foo
        out = [a, b, c]
      '
      to_match_snapshot


  it 'destructures recs', fn:
    expect
      fink2js "
        [{a, b}, c] = ni
        out = [a, b, c]
      "
      to_match_snapshot


  it 'destructures with defaults', fn:
    expect
      fink2js '
        [a=12, b=34] = foo
        out = [a, b]
      '
      to_match_snapshot

    expect
      fink2js '
        [[a, b=0]=[1, 2], c=34] = foo
        out = [a, b, c]
      '
      to_match_snapshot
