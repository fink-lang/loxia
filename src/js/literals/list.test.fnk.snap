// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list compiles 1`] = `
"const array1_0 = [];
const array2_0 = [1];
const array3_0 = [1, 2];
const array4_0 = [1,, 2];
const array5_0 = [a + 1 + 45 + b + c, [1, 2], (3 + 3) * 2];
export const array1 = array1_0,
      array2 = array2_0,
      array3 = array3_0,
      array4 = array4_0,
      array5 = array5_0;"
`;

exports[`list compiles as partial 1`] = `
"ˆpartial => [1, ...ˆpartial, 9];

ˆpartial => [ˆpartial, ˆpartial];"
`;

exports[`list compiles spread 1`] = `
"const bar_0 = [1, 2, ...b];
const spam_0 = [1, 2, ...b, 123];
export const foo = b,
      bar = bar_0,
      spam = spam_0;"
`;

exports[`list compiles spread 2`] = `
"const foo_0 = [1, 2, 3];
const bar_0 = [1, 2, 3, 4, 5, 6];
const spam_0 = [7];
const shrub_0 = [1, 2, 3, 4, 5, 6, 7];
const ni_0 = [...nu, na];
const ham_0 = [...nu, ...na];
export const foo = foo_0,
      bar = bar_0,
      spam = spam_0,
      shrub = shrub_0,
      ni = ni_0,
      ham = ham_0;"
`;

exports[`unpacking list destructures empty 1`] = `
"const out_0 = [foo[0], foo[2]];
export const out = out_0;"
`;

exports[`unpacking list destructures recs 1`] = `
"const drec_0 = ni[0];
const out_0 = [drec_0.a, drec_0.b, ni[1]];
export const out = out_0;"
`;

exports[`unpacking list destructures simple 1`] = `
"const item_3 = 1;
const item_2 = 2;
const foo_0 = [item_2, item_3];
const out_0 = [item_3, item_2, 3, foo_0, 1 + 2, item_2, item_3];
export const foo = foo_0,
      out = out_0;"
`;

exports[`unpacking list destructures spread 1`] = `
"const out_0 = [foo[0], [...foo.slice(1)].reverse()[0]];
export const out = out_0;"
`;

exports[`unpacking list destructures spread 2`] = `
"const rtail_0 = [...foo.slice(1)].reverse();
const out_0 = [foo[0], rtail_0[1], rtail_0[0]];
export const out = out_0;"
`;

exports[`unpacking list destructures spread 3`] = `
"const out_0 = [foo[0], ...foo.slice(1)];
export const out = out_0;"
`;

exports[`unpacking list destructures spread 4`] = `
"const rtail_0 = [...foo.slice(1)].reverse();
const out_0 = [foo[0], [...rtail_0.slice(1)].reverse(), rtail_0[0]];
export const out = out_0;"
`;

exports[`unpacking list destructures with defaults 1`] = `
"const hdm_0 = foo[0];
const hdm_1 = foo[1];
const out_0 = [undefined === hdm_0 ? 12 : hdm_0, undefined === hdm_1 ? 34 : hdm_1];
export const out = out_0;"
`;

exports[`unpacking list destructures with defaults 2`] = `
"const hdm_1 = foo[0];
const items_0 = undefined === hdm_1 ? [1, 2] : hdm_1;
const hdm_0 = items_0[1];
const hdm_2 = foo[1];
const out_0 = [items_0[0], undefined === hdm_0 ? 0 : hdm_0, undefined === hdm_2 ? 34 : hdm_2];
export const out = out_0;"
`;
